cbor2/encoder.py

```diff
@@ -114,6 +114,31 @@ def encode_map(encoder, value):
         encoder.encode(val)
 
 
+def encode_sortable_key(val):
+    '''Takes a key and calculates the length of it's optimal byte representation'''
+    if val.__class__ == bytes:
+        return len(val), val
+    if val.__class__ == unicode:
+        x = val.encode('utf-8')
+        return len(x), x
+    if val.__class__ == int:
+        bytelen = len(hex(abs(val))[1:]) // 2
+        return bytelen, val
+    else:
+        s = 'Canonical serialization requires string, bytes or integer keys in a mapping'
+        raise CBOREncodeError(s)
+
+
+@shareable_encoder
+def encode_canonical_map(encoder, value):
+    '''Reorder keys according to Canonical CBOR specification'''
+    keyed_keys = ((encode_sortable_key(key), key) for key in value.keys())
```

Comment:
Couldn't you have simply encoded the keys and then sorted the results? What do you need `encode_sortable_key()` for?

Reply:
I couldn't just call encoder.encode(key) since that wouldn't return a value, it would just write to the output file...

Reply:
Right, I keep forgetting that :) The RFC is a bit unclear on how this is supposed to work. Anyhow, I think I can improve on that integer encoding bit.

Reply:
i've been playing with this and just calling dumps(value, canonical=True) for every key seems the most straightforward.

Reply:
Yep, that would work for me.